rfcs/2022-03-15-11851-ingest-opentelemtry-traces.md

```diff
@@ -0,0 +1,384 @@
+  string appVersion = 10;
+}
+
+```
+
+Opentelemetry [trace format][otlp-proto-def] (condensed):
+
+```protobuf
+message InstrumentationLibrarySpans {
+  opentelemetry.proto.common.v1.InstrumentationLibrary instrumentation_library = 1;
+  repeated Span spans = 2;
+  string schema_url = 3;
+}
+
+message Span {
+  bytes trace_id = 1;
+  bytes span_id = 2;
+  string     = 3;
+  bytes parent_span_id = 4;
+  string name = 5;
+
+  enum SpanKind {
+    SPAN_KIND_UNSPECIFIED = 0;
+    SPAN_KIND_INTERNAL = 1;
+    SPAN_KIND_SERVER = 2;
+    SPAN_KIND_CLIENT = 3;
+    SPAN_KIND_PRODUCER = 4;
+    SPAN_KIND_CONSUMER = 5;
+  }
+
+  SpanKind kind = 6;
+  fixed64 start_time_unix_nano = 7;
+  fixed64 end_time_unix_nano = 8;
+  repeated opentelemetry.proto.common.v1.KeyValue attributes = 9;
+  uint32 dropped_attributes_count = 10;
+
+  message Event {
+    fixed64 time_unix_nano = 1;
+    string name = 2;
+    repeated opentelemetry.proto.common.v1.KeyValue attributes = 3;
+    uint32 dropped_attributes_count = 4;
+  }
+
+  repeated Event events = 11;
+  uint32 dropped_events_count = 12;
+
+  message Link {
+    bytes trace_id = 1;
+    bytes span_id = 2;
+    string trace_state = 3;
+    repeated opentelemetry.proto.common.v1.KeyValue attributes = 4;
+    uint32 dropped_attributes_count = 5;
+  }
+
+  repeated Link links = 13;
+  uint32 dropped_links_count = 14;
+  Status status = 15;
+}
+```
+
+The key construct in all trace formats is the **span** and traces are a set of spans. The OpenTelemetry span structure
+is rather verbose and comes with complex nested field. The [Datadog approach][span-conversion] is either to ignore those
+(e.g. the links field is ignored) or encode the complete field into a text representation (e.g. events are encoded using
+JSON) and include the resulting value into the tags (a.k.a Meta) map.
+
+This makes the opposite conversion a bit complicated if we want it to be completely symetrical but there was already an
+[attempt][otlp-dd-trace-receiver] allow Datadog traces ingestion in the OpenTelemetry collector. While this PR was
+closed unmerged this provide a valuable example. Anyways the [otlp-and-other-formats][OpenTelemetry] acknowledges that
+some of the OpenTelemetry contruct ends up being stored as tags or annotations in other formats.
+
+Anyway the OpenTelemtry to Datadog traces conversion is dictacted by existing implementation in both the `trace-agent`
+and the Datadog exporter as users will expect a consistent behaviour from one solution to another. The same
+consideration applies for APM stats computation, as [official implementations][apm-stats-computation] already provides a
+reference that define what should be done to get the same result with Vector in the loop. The other way, from Datadog to
+OpenTelemetry is less common as of today but while implementing conversions we should ensure that the following path is
+at best idempotent but at least fully functional:
```

Comment:
Do you have any specific concerns about the parts that won't be idempotent?

Reply:
I double checked both data structure, I don't find anything that would prevent the double conversions to be idempotent. I will expand a bit on that to clearly highlight that there is not blocking point and that the implementation will enforce that.

Reply:
Great!